apiserver/facades/controller/charmrevisionupdater/updater_test.go

```diff
@@ -4,136 +4,230 @@
+	resources.EXPECT().SetCharmStoreResources("app-1", expectedResources, gomock.Any()).Return(nil).AnyTimes()
+
+	client := mocks.NewMockCharmhubRefreshClient(ctrl)
+	matcher := charmhubConfigMatcher{expected: []charmhubConfigExpected{
+		{"charm-3", 1},
+	}}
+	client.EXPECT().Refresh(gomock.Any(), matcher).Return([]transport.RefreshResponse{
+		{
+			Entity: transport.RefreshEntity{
+				Resources: []transport.ResourceRevision{
+					{
+						Download: transport.ResourceDownload{
+							HashSHA384: "59e1748777448c69de6b800d7a33bbfb9ff1b463e44354c3553bcdb9c666fa90125a3c79f90397bdf5f6a13de828684f",
+							Size:       5,
+						},
+						Name:     "reza",
+						Revision: 7,
+						Type:     "file",
+					},
+					{
+						Download: transport.ResourceDownload{
+							HashSHA384: "03130092073c5ac523ecb21f548b9ad6e1387d1cb05f3cb892fcc26029d01428afbe74025b6c567b6564a3168a47179a",
+							Size:       6,
+						},
+						Name:     "rezb",
+						Revision: 1,
+						Type:     "file",
+					},
+				},
+				Revision: 1,
+			},
+			ID:   "charm-3",
+			Name: "resourcey",
+		},
+	}, nil)
+
+	state := makeState(c, ctrl, resources)
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "ch", "resourcey", "charm-3", "app-1", 1),
+	}, nil).AnyTimes()
+
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("ch:resourcey-1")).Return(nil)
+
+	newClient := func(st charmrevisionupdater.State, metadata map[string]string) (charmrevisionupdater.CharmhubRefreshClient, error) {
+		return client, nil
+	}
+	updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, nil, newClient)
+	c.Assert(err, jc.ErrorIsNil)
 
-	curl := charm.MustParseURL("cs:quantal/mysql")
-	_, err := s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+	result, err := updater.UpdateLatestRevisions()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Error, gc.IsNil)
+}
 
-	curl = charm.MustParseURL("cs:quantal/wordpress")
-	_, err = s.State.LatestPlaceholderCharm(curl)
-	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+func (s *updaterSuite) TestCharmhubNoUpdate(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	client := mocks.NewMockCharmhubRefreshClient(ctrl)
+	matcher := charmhubConfigMatcher{expected: []charmhubConfigExpected{
+		{"charm-2", 42},
+	}}
+	client.EXPECT().Refresh(gomock.Any(), matcher).Return([]transport.RefreshResponse{
+		{
+			Entity: transport.RefreshEntity{Revision: 42},
+			ID:     "charm-2",
+			Name:   "postgresql",
+		},
+	}, nil)
+
+	state := makeState(c, ctrl, nil)
+	state.EXPECT().AllApplications().Return([]charmrevisionupdater.Application{
+		makeApplication(ctrl, "ch", "postgresql", "charm-2", "app-2", 42),
+	}, nil).AnyTimes()
+	state.EXPECT().AddCharmPlaceholder(charm.MustParseURL("ch:postgresql-42")).Return(nil)
+
+	newClient := func(st charmrevisionupdater.State, metadata map[string]string) (charmrevisionupdater.CharmhubRefreshClient, error) {
```

Comment:
If you add client to the updaterSuite, you can do:

func (s *updaterSuite) newClient(_ charmrevisionupdater.State, _ map[string]string) (charmrevisionupdater.CharmhubRefreshClient, error) {
	return s.chClient, nil
}

and 

updater, err := charmrevisionupdater.NewCharmRevisionUpdaterAPIState(state, nil, s.newClient)

reduce code duplication.  Much of the juju go mock use, adds the mocks to the suite, to facilitate ease of use.

Reply:
Done (well, I don't want to move everything to fields on the struct, but I created a wrapper function which has the same effect of reducing the duplication).